File: pmengine/src/strategies/order_test.rs
//! Test strategy that places a single order and stops.

use crate::strategy::{Signal, Strategy, StrategyContext, Urgency};
use crate::position::Fill;
use rust_decimal_macros::dec;

pub struct OrderTest {
    id: String,
    tokens: Vec<String>,
    order_placed: bool,
}

impl OrderTest {
    pub fn new() -> Self {
        // Vermont Governor 2026 - active market
        Self {
            id: "order_test".to_string(),
            tokens: vec!["41583919731714354912849507182398941127545694257513505398713274521520484370640".to_string()],
            order_placed: false,
        }
    }
}

impl Default for OrderTest {
    fn default() -> Self {
        Self::new()
    }
}

impl Strategy for OrderTest {
    fn id(&self) -> &str {
        &self.id
    }

    fn subscriptions(&self) -> Vec<String> {
        self.tokens.clone()
    }

    fn on_tick(&mut self, _ctx: &StrategyContext) -> Vec<Signal> {
        // Only place one order
        if self.order_placed {
            return vec![Signal::Hold];
        }

        self.order_placed = true;
        let token_id = "41583919731714354912849507182398941127545694257513505398713274521520484370640".to_string();

        tracing::info!("Placing test order: BUY 5 @ $0.01");

        vec![Signal::Buy {
            token_id,
            price: dec!(0.01),
            size: dec!(5),
            urgency: Urgency::Low,
        }]
    }

    fn on_fill(&mut self, fill: &Fill) {
        tracing::info!(order_id = %fill.order_id, "Order filled!");
    }
}

File: pmstrat/pmstrat/strategies/order_test.py
"""Test strategy that places and immediately cancels an order."""

from decimal import Decimal
from pmstrat import strategy, Context, Signal, Buy, Cancel, Hold, Urgency


# A real token ID from Polymarket - "Will BTC be above $100k on Jan 31?"
# This is a liquid market that should have active trading
TOKEN_ID = "21742633143463906290569050155826241533067272736897614950488156847949938836455"


_order_placed = False

@strategy(
    name="order_test",
    tokens=[TOKEN_ID],
    tick_interval_ms=5000,
    transpilable=False,  # Uses global state - Python-only test strategy
)
def on_tick(ctx: Context) -> list[Signal]:
    """Place a small order once, then hold."""
    global _order_placed

    if _order_placed:
        return [Hold()]

    token_id = "21742633143463906290569050155826241533067272736897614950488156847949938836455"
    _order_placed = True

    return [
        Buy(
            token_id=token_id,
            price=Decimal("0.01"),
            size=Decimal("5"),
            urgency=Urgency.LOW,
        )
    ]
